modules/fedimint-lnv2-server/src/lib.rs

```diff
@@ -0,0 +1,668 @@
+                    OutgoingWitness::Claim(preimage) => {
+                        if contract.expiration <= self.consensus_block_count(dbtx).await {
+                            return Err(LightningInputError::Expired);
+                        }
+
+                        if !contract.verify_preimage(preimage) {
+                            return Err(LightningInputError::InvalidPreimage);
+                        }
+
+                        dbtx.insert_new_entry(&PreimageKey(*contract_id), preimage)
+                            .await;
+
+                        (contract.claim_pk, contract.amount)
+                    }
+                    OutgoingWitness::Refund => {
+                        if contract.expiration > self.consensus_block_count(dbtx).await {
+                            return Err(LightningInputError::NotExpired);
+                        }
+
+                        (contract.refund_pk, contract.amount)
+                    }
+                    OutgoingWitness::Cancel(forfeit_signature) => {
+                        if !contract.verify_forfeit_signature(forfeit_signature) {
+                            return Err(LightningInputError::InvalidForfeitSignature);
+                        }
+
+                        (contract.refund_pk, contract.amount)
+                    }
+                }
+            }
+            Witness::Incoming(contract_id, agg_decryption_key) => {
+                let contract = dbtx
+                    .remove_entry(&IncomingContractKey(*contract_id))
+                    .await
+                    .ok_or(LightningInputError::UnknownContract)?;
+
+                if !contract
+                    .verify_agg_decryption_key(&self.cfg.consensus.tpe_agg_pk, agg_decryption_key)
+                {
+                    return Err(LightningInputError::InvalidDecryptionKey);
+                }
+
+                match contract.decrypt_preimage(agg_decryption_key) {
+                    Some(..) => (contract.commitment.claim_pk, contract.commitment.amount),
+                    None => (contract.commitment.refund_pk, contract.commitment.amount),
+                }
+            }
+        };
+
+        Ok(InputMeta {
+            amount: TransactionItemAmount {
+                amount,
+                fee: self.cfg.consensus.fee_consensus.contract_input,
+            },
+            pub_key,
+        })
+    }
+
+    async fn process_output<'a, 'b>(
+        &'a self,
+        dbtx: &mut DatabaseTransaction<'b>,
+        output: &'a LightningOutput,
+        out_point: OutPoint,
+    ) -> Result<TransactionItemAmount, LightningOutputError> {
+        let outcome = match output {
+            LightningOutput::Outgoing(contract) => {
+                if dbtx
+                    .insert_entry(&OutgoingContractKey(contract.contract_id()), contract)
+                    .await
+                    .is_some()
+                {
+                    return Err(LightningOutputError::ContractAlreadyExists);
```

Comment:
As mentioned above, lets add the `contract_id` for debugging purposes here and below